### DISCUSSION TITLE
Value Object não deveria ser imutável?

### DISCUSSION THREAD

**Discussion by oGrandeK** [balta-io/forum#384] [Tira Dúvidas] [answered]
Olá professor, participei de muitas aulas do senhor e o senhor usa bastante Value Objects para algumas propriedades das suas entidades, eu acho o conceito de VO animal mas me surgiu uma dúvida, todo lugar que eu procuro algo sobre VO, sempre me apontam que os VOs devem ser imutáveis e no seu caso o senhor sempre acaba fazendo os VOs para propriedades que em teoria podem ter a possibilidade de ser mutável. No caso da entidade User ter um Email, esse email hora ou outra o usuário pode querer alterar tanto que o senhor faz métodos para permitir essa alteração.

Enfim, Value Object não deveria ser imutável?

**Comment by andrebaltieri** [comment, marked as answer]
Bom dia, @oGrandeK, como vai?

Acredito que respondi esta pergunta em outra thread aqui, mas não consigo encontrar no momento!

A resposta curta e direta é sim! Eles devem ser imutáveis.
Porém, eu não sigo isto tão ao pé da letra, se eu vejo um cenário onde posso me beneficiar de uma alteração, eu faço, só privo para que somente meu VO possa alterar seu estado.

Se você olhar os novos cursos, estou com uma abordagem mais rígida, mas isto varia de acordo com cada cenário.
Flexibilidade, balanço e equilíbrio são o que sempre levo em conta.
